File: appMain/src/main/java/com/cartlc/tracker/data/DataStates.java
package com.cartlc.tracker.data;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by dug on 5/11/17.
 */

public class DataStates {

    public static class State {
        String full;
        String abbr;

        State(String full, String abbr) {
            this.full = full;
            this.abbr = abbr;
        }
    }

    static final State[] STATES = {
            new State("Alabama", "AL"),
            new State("Alaska", "AK"),
            new State("Arizona", "AZ"),
            new State("Arkansas", "AR"),
            new State("California", "CA"),
            new State("Colorado", "CO"),
            new State("Connecticut", "CT"),
            new State("Delaware", "DE"),
            new State("Florida", "FL"),
            new State("Georgia", "GA"),
            new State("Hawaii", "HI"),
            new State("Idaho", "ID"),
            new State("Illinois", "IL"),
            new State("Indiana", "IN"),
            new State("Iowa", "IA"),
            new State("Kansas", "KS"),
            new State("Kentucky", "KY"),
            new State("Louisiana", "LA"),
            new State("Maine", "ME"),
            new State("Maryland", "MD"),
            new State("Massachusetts", "MA"),
            new State("Michigan", "MI"),
            new State("Minnesota", "MN"),
            new State("Mississippi", "MS"),
            new State("Missouri", "MO"),
            new State("Montana", "MT"),
            new State("Nebraska", "NE"),
            new State("Nevada", "NV"),
            new State("New Hampshire", "NH"),
            new State("New Jersey", "NJ"),
            new State("New Mexico", "NM"),
            new State("New York", "NY"),
            new State("North Carolina", "NC"),
            new State("North Dakota", "ND"),
            new State("Ohio", "OH"),
            new State("Oklahoma", "OK"),
            new State( "Oregon", "OR"),
            new State("Pennsylvania", "PA"),
            new State("Rhode Island", "RI"),
            new State("South Carolina", "SC"),
            new State("South Dakota", "SD"),
            new State("Tennessee", "TN"),
            new State("Texas", "TX"),
            new State("Utah", "UT"),
            new State("Vermont", "VT"),
            new State("Virginia", "VA"),
            new State("Washington", "WA"),
            new State("West Virginia", "WV"),
            new State("Wisconsin", "WI"),
            new State("Wyoming", "WY"),
    };

    public static List<String> getUnusedStates(List<String> usedStates) {
        List<String> unused = new ArrayList();
        for (State state : STATES) {
            if (!usedStates.contains(state.full)) {
                unused.add(state.full);
            }
        }
        Collections.sort(unused);
        return unused;
    }

    public static boolean isValid(String scan) {
        return get(scan) != null;
    }

    public static State get(String scan) {
        for (State state : STATES) {
            if (state.full.equalsIgnoreCase(scan) || state.abbr.equalsIgnoreCase(scan)) {
                return state;
            }
        }
        return null;
    }

    public static String getAbbr(String scan) {
        State state = get(scan);
        if (state != null) {
            return state.abbr;
        }
        return null;
    }

    public static String getFull(String scan) {
        State state = get(scan);
        if (state != null) {
            return state.full;
        }
        return null;
    }
}

File: appMain/src/main/java/com/cartlc/tracker/model/data/DataStates.kt
/**
 * Copyright 2018, FleetTLC. All rights reserved
 */
package com.cartlc.tracker.model.data

/**
 * Created by dug on 5/11/17.
 */

object DataStates {

    internal val STATES = arrayOf(
            State("Alabama", "AL"),
            State("Alaska", "AK"),
            State("Arizona", "AZ"),
            State("Arkansas", "AR"),
            State("California", "CA"),
            State("Colorado", "CO"),
            State("Connecticut", "CT"),
            State("Delaware", "DE"),
            State("Florida", "FL"),
            State("Georgia", "GA"),
            State("Hawaii", "HI"),
            State("Idaho", "ID"),
            State("Illinois", "IL"),
            State("Indiana", "IN"),
            State("Iowa", "IA"),
            State("Kansas", "KS"),
            State("Kentucky", "KY"),
            State("Louisiana", "LA"),
            State("Maine", "ME"),
            State("Maryland", "MD"),
            State("Massachusetts", "MA"),
            State("Michigan", "MI"),
            State("Minnesota", "MN"),
            State("Mississippi", "MS"),
            State("Missouri", "MO"),
            State("Montana", "MT"),
            State("Nebraska", "NE"),
            State("Nevada", "NV"),
            State("New Hampshire", "NH"),
            State("New Jersey", "NJ"),
            State("New Mexico", "NM"),
            State("New York", "NY"),
            State("North Carolina", "NC"),
            State("North Dakota", "ND"),
            State("Ohio", "OH"),
            State("Oklahoma", "OK"),
            State("Oregon", "OR"),
            State("Pennsylvania", "PA"),
            State("Rhode Island", "RI"),
            State("South Carolina", "SC"),
            State("South Dakota", "SD"),
            State("Tennessee", "TN"),
            State("Texas", "TX"),
            State("Utah", "UT"),
            State("Vermont", "VT"),
            State("Virginia", "VA"),
            State("Washington", "WA"),
            State("West Virginia", "WV"),
            State("Wisconsin", "WI"),
            State("Wyoming", "WY"))

    class State internal constructor(var full: String, var abbr: String)

    fun getUnusedStates(usedStates: List<String>): List<String> {
        val unused = mutableListOf<String>()
        for (state in STATES) {
            if (!usedStates.contains(state.full) && !usedStates.contains(state.abbr)) {
                unused.add(state.full)
            }
        }
        unused.sort()
        return unused
    }

    fun isValid(scan: String?): Boolean {
        return scan?.let { get(it) != null } ?: false
    }

    operator fun get(scan: String): State? {
        for (state in STATES) {
            if (state.full.equals(scan, ignoreCase = true) || state.abbr.equals(scan, ignoreCase = true)) {
                return state
            }
        }
        return null
    }

    fun getAbbr(scan: String): String? {
        val state = get(scan)
        return state?.abbr
    }

    fun getFull(scan: String): String? {
        val state = get(scan)
        return state?.full
    }
}
